### PR TITLE
[FLINK-11910] [Yarn] add customizable yarn application type

### PR DESCRIPTION
**Pull request by HuangZhenQiu**
## What is the purpose of the change

Let user customize yarn application type tag by using dynamic properties.

## Brief change log
   - use flink-version as a default key for user to specify the flink version
   - If flink-version is set, the value will be added after original "Apache Flink ".

## Verifying this change

This change is verified in YARNSessionCapacitySchedulerITCase

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-yarn-tests/src/test/java/org/apache/flink/yarn/YARNSessionFIFOITCase.java [hunk 1]
```diff
@@ -110,6 +110,8 @@ public void testDetachedMode() throws InterruptedException, IOException {
 		args.add("-tm");
 		args.add("1024m");
 
+		args.add("-Dflink-version=1.6");
```
Comment (klion26) [inline flink-yarn-tests/src/test/java/org/apache/flink/yarn/YARNSessionFIFOITCase.java]: Maybe we should add assert logic somewhere?

File: flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java [hunk 2]
```diff
@@ -468,6 +470,8 @@ private void validateClusterSpecification(ClusterSpecification clusterSpecificat
 			flinkConfiguration.setString(dynProperty.getKey(), dynProperty.getValue());
 		}
 
+		this.flinkVersion = dynProperties.getOrDefault("flink-version", "");
```
Comment (klion26) [inline flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java]: Whether we could get the version from the properties already in this class, like `flinkJarPath`?

Comment (HuangZhenQiu) [reply flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java]: For most of the cases, we can get the precise version of flinkJarPath. But for some deployment systems that submit jobs from customers, different jar will be used. Probably, using dynamic properties can give more flexibility here. How do you think?

Comment (klion26) [reply flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java]: Do you want to show the same version for different jars?

Comment (HuangZhenQiu) [reply flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java]: No, different versions can be specified by the dynamic properties right? It can be either through cmd or the backend service that launches flink job.

**Comment by suez1224** [comment]
@flinkbot approve description

**Comment by suez1224** [comment]
Instead of customize just the version, can we just let user override the entire application type, which is more flexible?

**Comment by rmetzger** [comment]
@flinkbot approve description consensus

**Comment by rmetzger** [comment]
@suez1224 I'm sorry that the @flinkbot didn't know that you are a committer of Flink.
I've updated the config and restarted the bot. It should soon show in the tracking comment that you are a committer!

Regarding the PR: I agree that we should make the application type overridable entirely, not just the version! (this is also more intuitive from a configuration point of view)

**Comment by rmetzger** [comment]
Thank you for addressing our comments.

@flinkbot attention @GJL

**Comment by HuangZhenQiu** [comment]
@rmetzger 
It looks there are some random test failures after rebase master. Any suggestion for this situation?

**Comment by rmetzger** [comment]
I've restarted the failed job, let's see

**Comment by HuangZhenQiu** [comment]
@rmetzger 
I saw a test module is cancelled. Should I look deeper into it?

**Comment by rmetzger** [comment]
I guess this was only a glitch. Travis is all green now.

**Comment by HuangZhenQiu** [comment]
@suez1224 @rmetzger 
Would you please have one more round of review?

File: flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java [hunk 3]
```diff
@@ -988,7 +992,7 @@ public ApplicationReport startAppMaster(
 		final String customApplicationName = customName != null ? customName : applicationName;
 
 		appContext.setApplicationName(customApplicationName);
-		appContext.setApplicationType("Apache Flink");
+		appContext.setApplicationType(applicationType.isEmpty() ? "Apache Flink" : applicationType);
```
Comment (suez1224) [inline flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java]: Can't we just do the following?
  appContext.setApplicationType(dynProperties.getOrDefault("application-type",  "Apache Flink"));

File: flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java [hunk 4]
```diff
@@ -138,6 +138,8 @@
 	private String zookeeperNamespace;
 
 	private String nodeLabel;
+	/* This field can be override by dynamic property application-type */
+	private String applicationType;
```
Comment (suez1224) [inline flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java]: We don't need this instance variable.

File: flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java [hunk 5]
```diff
@@ -468,6 +470,8 @@ private void validateClusterSpecification(ClusterSpecification clusterSpecificat
 			flinkConfiguration.setString(dynProperty.getKey(), dynProperty.getValue());
 		}
 
+		this.applicationType = dynProperties.getOrDefault("application-type", "");
```
Comment (suez1224) [inline flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java]: We dont need this line as well..

**Comment by suez1224** [comment]
@flinkbot approve architecture

**Comment by GJL** [comment]
> Instead of customize just the version, can we just let user override the entire application type, which is more flexible?

The default could still be to include the version in the application type.

File: flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (HuangZhenQiu) [reply flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java]: @GJL Agree. I will change it to using command line option.

**Comment by HuangZhenQiu** [comment]
@GJL Would you please have a look the change?

**Comment by GJL** [comment]
@HuangZhenQiu Thanks for the update. I will try to take a look later today.

File: flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java [hunk 6]
```diff
@@ -1285,6 +1287,13 @@ public void setName(String name) {
 		customName = name;
 	}
 
+	public void setApplicationType(String type) {
+		if (type == null) {
```
Comment (GJL) [inline flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java]: Let's use `org.apache.flink.util.Preconditions`:

`this.applicationType = Preconditions.checkNotNull(applicationType, "applicationType must not be null");`

`setName` should also be refactored in a separate commit.

File: flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java [hunk 7]
```diff
@@ -201,6 +202,7 @@ public FlinkYarnSessionCli(
 			.build();
 		streaming = new Option(shortPrefix + "st", longPrefix + "streaming", false, "Start Flink in streaming mode");
 		name = new Option(shortPrefix + "nm", longPrefix + "name", true, "Set a custom name for the application on YARN");
+		applicationType = new Option(shortPrefix + "at", longPrefix + "applicationType", true, "Set a custom application type for the application on YARN");
```
Comment (GJL) [inline flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: `yarn_setup.md` should be updated as well I think. Specifically this part:

```
{% highlight bash %}
Usage:
   Optional
     -D <arg>                        Dynamic properties
     -d,--detached                   Start detached
     -jm,--jobManagerMemory <arg>    Memory for JobManager Container with optional unit (default: MB)
     -nm,--name                      Set a custom name for the application on YARN
     -q,--query                      Display available YARN resources (memory, cores)
     -qu,--queue <arg>               Specify YARN queue.
     -s,--slots <arg>                Number of slots per TaskManager
     -tm,--taskManagerMemory <arg>   Memory per TaskManager Container with optional unit (default: MB)
     -z,--zookeeperNamespace <arg>   Namespace to create the Zookeeper sub-paths for HA mode
{% endhighlight %}
```

**Comment by rmetzger** [comment]
Once @GJL's comments have been resolved, I'm +1 on merging this.

**Comment by HuangZhenQiu** [comment]
@GJL @rmetzger 
I changed the PR according to the comments of Gary. Please have another review at your most convenient time.

**Comment by rmetzger** [comment]
@flinkbot approve all

**Comment by rmetzger** [comment]
I will merge this change now

**Comment by GJL** [comment]
Thanks for merging @rmetzger
